Enable spi_loopback tests on nucleo_h745zi_q boards

bene42 opened the pull request:
The SIP loopback test have no configuration and overlay for the Nucleo-H745ZI-Q board.
Profide a configuration and overlay for these board.

bene42 committed:
tests: drivers: spi loopback testing on the nucleo_h745zi_q boards

Connect D12(PA6) & D22(PB5) pins on the board to pass the test.

Signed-off-by: Benedikt Spranger <b.spranger@linutronix.de>
Reviewed-by: Vasilij Strassheim <vasilij.strassheim@linutronix.de>

bene42 commented:
A gentle ping.

erwango reviewed (changes requested):
Not totally in line with the change, the test won't be run automatically by twister as the board misses to declare spi support in board yaml file (and misses to declare SPI nodes). Also test yaml file should allow test on this board (in platform allow list).

bene42 commented on tests/drivers/spi/spi_loopback/boards/nucleo_h745zi_q_stm32h745xx_m7.overlay:
```diff
@@ -0,0 +1,50 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/dt-bindings/memory-attr/memory-attr-arm.h>
```
Thanks for the remark - I look into that.

bene42 committed:
tests: drivers: Remove unnessesary include

The nucleo_h7X3zi overlays use memory attributes and therfore the
dedicated include file. This is unnessesary, since the memory attribute
include is already included from the underlaying CPU base file
"stm32h7.dtsi". Remove the unnessesary includes.

Signed-off-by: Benedikt Spranger <b.spranger@linutronix.de>

bene42 committed:
boards: st: nucleo_h745zi_q: Wire up SPI tests

The previously added overlay can now be used. Wire up SPI tests for the
Nucleo h745zi-q board.

Signed-off-by: Benedikt Spranger <b.spranger@linutronix.de>

bene42 commented:
## Results

### M4

    $ ./scripts/twister --device-testing --device-serial /dev/ttyACM0 --device-serial-baud 115200 -p nucleo_h745zi_q/stm32h745xx/m4  -T tests/drivers/spi/spi_loopback -X spi_loopback
    INFO    - Using Ninja..
    INFO    - Zephyr version: v3.7.0-798-g5c6c5468389e
    INFO    - Using 'zephyr' toolchain.
    INFO    - Building initial testsuite list...
    INFO    - Writing JSON report /home/bene/work/zephyr/zephyr/twister-out/testplan.json
    
    Device testing on:
    
    | Platform                       | ID   | Serial device   |
    |--------------------------------|------|-----------------|
    | nucleo_h745zi_q/stm32h745xx/m4 |      | /dev/ttyACM0    |
    
    INFO    - JOBS: 4
    INFO    - Adding tasks to the queue...
    INFO    - Added initial list of jobs to queue
    INFO    - Total complete:   19/  23  82%  skipped:   19, failed:    0, error:   INFO    - Total complete:   20/  23  86%  skipped:   20, failed:    0, error:   INFO    - Total complete:   21/  23  91%  skipped:   21, failed:    0, error:   INFO    - Total complete:   22/  23  95%  skipped:   22, failed:    0, error:   INFO    - Total complete:   23/  23  100%  skipped:   22, failed:    0, error:    0
    INFO    - 23 test scenarios (23 test instances) selected, 22 configurations skipped (18 by static filter, 4 at runtime).
    INFO    - 1 of 23 test configurations passed (100.00%), 0 failed, 0 errored, 22 skipped with 0 warnings in 35.37 seconds
    INFO    - In total 1 test cases were executed, 44 skipped on 1 out of total 2 platforms (50.00%)
    INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.
    
    Hardware distribution summary:
    
    | Board                          | ID   |   Counter |
    |--------------------------------|------|-----------|
    | nucleo_h745zi_q/stm32h745xx/m4 |      |         1 |
    INFO    - Saving reports...
    INFO    - Writing JSON report /home/bene/work/zephyr/zephyr/twister-out/twister.json
    INFO    - Writing xunit report /home/bene/work/zephyr/zephyr/twister-out/twister.xml...
    INFO    - Writing xunit report /home/bene/work/zephyr/zephyr/twister-out/twister_report.xml...
    INFO    - Run completed

### M7

    $ ./scripts/twister --device-testing --device-serial /dev/ttyACM0 --device-serial-baud 115200 -p nucleo_h745zi_q/stm32h745xx/m7  -T tests/drivers/spi/spi_loopback -X spi_loopback
    Renaming output directory to /home/bene/work/zephyr/zephyr/twister-out.1
    INFO    - Using Ninja..
    INFO    - Zephyr version: v3.7.0-798-g5c6c5468389e
    INFO    - Using 'zephyr' toolchain.
    INFO    - Building initial testsuite list...
    INFO    - Writing JSON report /home/bene/work/zephyr/zephyr/twister-out/testplan.json
    
    Device testing on:
    
    | Platform                       | ID   | Serial device   |
    |--------------------------------|------|-----------------|
    | nucleo_h745zi_q/stm32h745xx/m7 |      | /dev/ttyACM0    |
    
    INFO    - JOBS: 4
    INFO    - Adding tasks to the queue...
    INFO    - Added initial list of jobs to queue
    INFO    - Total complete:   19/  23  82%  skipped:   19, failed:    0, error:   INFO    - Total complete:   20/  23  86%  skipped:   20, failed:    0, error:   INFO    - Total complete:   21/  23  91%  skipped:   21, failed:    0, error:   INFO    - Total complete:   22/  23  95%  skipped:   22, failed:    0, error:   INFO    - Total complete:   23/  23  100%  skipped:   22, failed:    0, error:    0
    INFO    - 23 test scenarios (23 test instances) selected, 22 configurations skipped (18 by static filter, 4 at runtime).
    INFO    - 1 of 23 test configurations passed (100.00%), 0 failed, 0 errored, 22 skipped with 0 warnings in 39.77 seconds
    INFO    - In total 1 test cases were executed, 44 skipped on 1 out of total 2 platforms (50.00%)
    INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.
    
    Hardware distribution summary:
    
    | Board                          | ID   |   Counter |
    |--------------------------------|------|-----------|
    | nucleo_h745zi_q/stm32h745xx/m7 |      |         1 |
    INFO    - Saving reports...
    INFO    - Writing JSON report /home/bene/work/zephyr/zephyr/twister-out/twister.json
    INFO    - Writing xunit report /home/bene/work/zephyr/zephyr/twister-out/twister.xml...
    INFO    - Writing xunit report /home/bene/work/zephyr/zephyr/twister-out/twister_report.xml...
    INFO    - Run completed

erwango reviewed (changes requested):
Thanks for the changes but that's not fully okay.

When declaring `spi` at board level, this info that board could be used for all tests and samples that require spi to build and run. However you enable spi node only in spi_loopback test.
What you could do is, as done on several other boards, to declare spi node at board level (status okay + pins + cs-gpios) and add test specifics in board overlays you created (dma, cache, speed config)).

bene42 commented:
> Thanks for the changes but that's not fully okay.
What a bummer! :)))

> When declaring `spi` at board level, this info that board could be used for all tests and samples that require spi to
> build and run.
That was mainly my intention.

> However you enable spi node only in spi_loopback test.
OK, that was unintentional. I did a complete twister run for that board and...
...missread the results. OK. Need to change that.

> What you could do is, as done on several other boards, to declare spi node at board level 
> (status okay + pins + cs gpios)
>  and add test specifics in board overlays you created (dma, cache, speed config)).
To be honest: I dislike that approach. These boards have pin sockets and headers free to use. I have a wide and colorful set of shields using different pinmux and functions for these pins. Since these pins are "no connection" by default it suggest itself to leave them "disabled" in the device tree.

But since I don't see any real alternative to that, I will do it that way.

Thanks for your review.

erwango commented:
> To be honest: I dislike that approach. These boards have pin sockets and headers free to use. I have a wide and colorful set of shields using different pinmux and functions for these pins. Since these pins are "no connection" by default it suggest itself to leave them "disabled" in the device tree.

I agree the solution we're enforcing is not 100% satisfying, though that's the best way we've found to enable boards to be compatible with most samples/tests without having (or at least: while minimizing the need) to provide a new overlay for each test (like enabling nodes in each samples/tests). This default configuration is provided in order to demonstrate board capabilities and enable compatibility with tests and samples at a minimum cost.
This is done under consideration that upstream boards configurations are made for demo and CI testing of upstream drivers and SoCs.
Then, should you have out-of-tree needs, you'll need to maintain your configurations out of tree (as community of users will have diverging specific needs which are hard to conciliate upstream).

bene42 committed:
boards: st: nucleo_h745zi_q: Enable SPI on Arduino extention

The board is capable to serve SPI on the Arduino extention header.
Configure and enable SPI to work on the extention.

Signed-off-by: Benedikt Spranger <b.spranger@linutronix.de>

bene42 commented:
Sorry for the delay, but it costs me some time to figure out the extraordinary configuration and use of fixtures... :(

But back on track:

### available SPI tests

I looked through the zephyr tree for SPI tests usable for the Nucleo h745zi-q board. Here is a commented list of available tests:

|                         SPI tests                         |     comment     |
|-----------------------------------------------------------|-----------------|
| tests/boards/altera_max10/qspi/testcase.yaml              | incompatible hw |
| tests/boards/espressif_esp32/cache_coex/testcase.yaml     | incompatible hw |
| tests/boards/mec15xxevb_assy6853/qspi/testcase.yaml       | incompatible hw |
| tests/boards/mec172xevb_assy6906/qspi/testcase.yaml       | incompatible hw |
| tests/drivers/espi/testcase.yaml                          | incompatible hw |
| tests/drivers/flash/common/testcase.yaml                  | incompatible hw |
| tests/drivers/mspi/api/testcase.yaml                      | incompatible hw |
| tests/drivers/mspi/flash/testcase.yaml                    | incompatible hw |
|-----------------------------------------------------------|-----------------|
| tests/drivers/build_all/can/testcase.yaml                 |   build only    |
| tests/drivers/build_all/eeprom/testcase.yaml              |   build only    |
| tests/drivers/build_all/ethernet/testcase.yaml            |   build only    |
| tests/drivers/build_all/gpio/testcase.yaml                |   build only    |
| tests/drivers/build_all/led_strip/testcase.yaml           |   build only    |
| tests/drivers/build_all/rtc/testcase.yaml                 |   build only    |
|-----------------------------------------------------------|-----------------|
| tests/drivers/display/display_read_write/testcase.yaml    | extra hw needed |
| tests/drivers/spi/dt_spec/testcase.yaml                   | see extra rem.  |
| tests/drivers/spi/spi_controller_peripheral/testcase.yaml | unavail 2. SPI  |
| tests/drivers/spi/spi_error_cases/testcase.yaml           | unavail 2. SPI  |
| tests/drivers/spi/spi_loopback/testcase.yaml              | works           |

#### Test dt_spec
I was not able to run the test without error

    ERROR   - nucleo_h745zi_q/stm32h745xx/m4 tests/drivers/spi/dt_spec/drivers.spi.dt_spec 
    FAILED : Testsuite mismatch

I am not shure, but it seems that a "build only" hint is missing here.

### Test results

#### M4
    INFO    - Using Ninja..
    INFO    - Zephyr version: v3.7.0-1196-gfc7b1bb5fdc5
    INFO    - Using 'zephyr' toolchain.
    INFO    - Building initial testsuite list...
    INFO    - Writing JSON report /home/bene/work/zephyr/testing/zephyr/twister-out/testplan.json
    
    Device testing on:
    
    | Platform                       |                       ID | Serial device   |
    |--------------------------------|--------------------------|-----------------|
    | nucleo_h745zi_q/stm32h745xx/m4 | 004600283033510435393935 | /dev/ttyACM0    |
    
    INFO    - JOBS: 4
    INFO    - Adding tasks to the queue...
    INFO    - Added initial list of jobs to queue
    INFO    - Total complete:   13/  23  56%  skipped:   13, failed:    0, error:    0
    INFO    - Total complete:   14/  23  60%  skipped:   14, failed:    0, error:    0
    INFO    - Total complete:   15/  23  65%  skipped:   14, failed:    0, error:    0
    INFO    - Total complete:   16/  23  69%  skipped:   14, failed:    0, error:    0
    INFO    - Total complete:   17/  23  73%  skipped:   14, failed:    0, error:    0
    INFO    - Total complete:   18/  23  78%  skipped:   15, failed:    0, error:    0
    INFO    - Total complete:   19/  23  82%  skipped:   16, failed:    0, error:    0
    INFO    - Total complete:   20/  23  86%  skipped:   16, failed:    0, error:    0
    INFO    - Total complete:   21/  23  91%  skipped:   17, failed:    0, error:    0
    INFO    - Total complete:   22/  23  95%  skipped:   18, failed:    0, error:    0
    INFO    - Total complete:   23/  23  100%  skipped:   18, failed:    0, error:    0
    INFO    - 23 test scenarios (23 test instances) selected, 18 configurations skipped (12 by static filter, 6 at runtime).
    INFO    - 5 of 23 test configurations passed (100.00%), 0 failed, 0 errored, 18 skipped with 0 warnings in 128.87 seconds
    INFO    - In total 5 test cases were executed, 36 skipped on 1 out of total 2 platforms (50.00%)
    INFO    - 5 test configurations executed on platforms, 0 test configurations were only built.
    
    Hardware distribution summary:
    
    | Board                          |                       ID |   Counter |   Failures |
    |--------------------------------|--------------------------|-----------|------------|
    | nucleo_h745zi_q/stm32h745xx/m4 | 004600283033510435393935 |         5 |          0 |
    
    INFO    - Saving reports...
    INFO    - Writing JSON report /home/bene/work/zephyr/testing/zephyr/twister-out/twister.json
    INFO    - Writing xunit report /home/bene/work/zephyr/testing/zephyr/twister-out/twister.xml...
    INFO    - Writing xunit report /home/bene/work/zephyr/testing/zephyr/twister-out/twister_report.xml...
    INFO    - Run completed

#### M7
    INFO    - Using Ninja..
    INFO    - Zephyr version: v3.7.0-1196-gfc7b1bb5fdc5
    INFO    - Using 'zephyr' toolchain.
    INFO    - Building initial testsuite list...
    INFO    - Writing JSON report /home/bene/work/zephyr/testing/zephyr/twister-out/testplan.json
    
    Device testing on:
    
    | Platform                       |                       ID | Serial device   |
    |--------------------------------|--------------------------|-----------------|
    | nucleo_h745zi_q/stm32h745xx/m7 | 004600283033510435393935 | /dev/ttyACM0    |
    
    INFO    - JOBS: 4
    INFO    - Adding tasks to the queue...
    INFO    - Added initial list of jobs to queue
    INFO    - Total complete:   13/  23  56%  skipped:   12, failed:    0, error:    0
    INFO    - Total complete:   14/  23  60%  skipped:   12, failed:    0, error:    0
    INFO    - Total complete:   15/  23  65%  skipped:   12, failed:    0, error:    0
    INFO    - Total complete:   16/  23  69%  skipped:   12, failed:    0, error:    0
    INFO    - Total complete:   17/  23  73%  skipped:   13, failed:    0, error:    0
    INFO    - Total complete:   18/  23  78%  skipped:   14, failed:    0, error:    0
    INFO    - Total complete:   19/  23  82%  skipped:   15, failed:    0, error:    0
    INFO    - Total complete:   20/  23  86%  skipped:   16, failed:    0, error:    0
    INFO    - Total complete:   21/  23  91%  skipped:   16, failed:    0, error:    0
    INFO    - Total complete:   22/  23  95%  skipped:   16, failed:    0, error:    0
    INFO    - Total complete:   23/  23  100%  skipped:   16, failed:    0, error:    0
    
    INFO    - 23 test scenarios (23 test instances) selected, 16 configurations skipped (12 by static filter, 4 at runtime).
    INFO    - 7 of 23 test configurations passed (100.00%), 0 failed, 0 errored, 16 skipped with 0 warnings in 183.64 seconds
    INFO    - In total 7 test cases were executed, 32 skipped on 1 out of total 2 platforms (50.00%)
    INFO    - 7 test configurations executed on platforms, 0 test configurations were only built.
    
    Hardware distribution summary:
    
    | Board                          |                       ID |   Counter |   Failures |
    |--------------------------------|--------------------------|-----------|------------|
    | nucleo_h745zi_q/stm32h745xx/m7 | 004600283033510435393935 |         7 |          0 |
    
    INFO    - Saving reports...
    INFO    - Writing JSON report /home/bene/work/zephyr/testing/zephyr/twister-out/twister.json
    INFO    - Writing xunit report /home/bene/work/zephyr/testing/zephyr/twister-out/twister.xml...
    INFO    - Writing xunit report /home/bene/work/zephyr/testing/zephyr/twister-out/twister_report.xml...
    INFO    - Run completed

erwango commented:
> Sorry for the delay, but it costs me some time to figure out the extraordinary configuration and use of fixtures... :(

I agree this could be a long journey. Don't hesitate to submit changes to the documentation if you find pieces are missing.

bene42 commented:
A gentle ping.
